l1-contracts/deploy-scripts/PrepareZKChainRegistrationCalldata.s.sol

```diff
@@ -0,0 +1,363 @@
+        config.baseToken = toml.readAddress("$.chain.base_token_addr");
+        config.governance = Bridgehub(config.bridgehub).owner();
+
+        bytecodes.l2SharedBridgeBytecode = Utils.readHardhatBytecode(
+            "/../l2-contracts/artifacts-zk/contracts/bridge/L2SharedBridge.sol/L2SharedBridge.json"
+        );
+
+        bytecodes.l2SharedBridgeProxyBytecode = Utils.readHardhatBytecode(
+            "/../l2-contracts/artifacts-zk/@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol/TransparentUpgradeableProxy.json"
+        );
+        bytecodes.beaconProxy = Utils.readHardhatBytecode(
+            "/../l2-contracts/artifacts-zk/@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol/BeaconProxy.json"
+        );
+    }
+
+    function checkBaseTokenAddress() internal view {
+        if (config.baseToken == address(0)) {
+            revert("Base token address is not set");
+        }
+
+        // Check if it's ethereum address
+        if (config.baseToken == ADDRESS_ONE) {
+            return;
+        }
+
+        // Does not prevent registering a non-ERC-20 contract as a token
+        // But calling the ERC-20 methods here on non-ERC-20 will fail without a readable revert message
+        if (config.baseToken.code.length == 0) {
+            revert("Token address is an EOA");
+        }
+
+        console.log("Using base token address:", config.baseToken);
+    }
+
+    function prepareRegisterBaseTokenCall() internal view returns (IGovernance.Call memory) {
+        Bridgehub bridgehub = Bridgehub(config.bridgehub);
+
+        bytes memory data = abi.encodeCall(bridgehub.addToken, (config.baseToken));
+
+        return IGovernance.Call({target: config.bridgehub, value: 0, data: data});
+    }
+
+    // @dev Computes the address of the L2 bridge and the L2 bridge proxy
+    // The variables that are used to compute the address are:
+    // - Salt
+    // - L2 bridge impl bytecode hash
+    // - Era chain ID
+    // - Beacon proxy bytecode hash
+    // - L1 governance address(owner of the L2 bridge)
+    // - L1 ERC20 bridge proxy address
+    // - Default ERC20 proxy address
+    // - Deployer's address
+    function computeL2BridgeAddress() internal view returns (address) {
+        bytes32 salt = "";
+        bytes32 bridgeBytecodeHash = L2ContractHelper.hashL2Bytecode(bytecodes.l2SharedBridgeBytecode);
+        bytes memory bridgeConstructorData = abi.encode(config.eraChainId);
+
+        address deployer;
+        address l2GovernanceAddress;
+
+        // ZKsync's protocol design assumes that the addresses of all the smart contracts that are sending L1->L2
+        // messages are aliased. We have to check if the sender is an EOA and apply the alias if it is not.
+        if (isEOA(msg.sender)) {
+            deployer = msg.sender;
+        } else {
+            deployer = AddressAliasHelper.applyL1ToL2Alias(msg.sender);
```

Comment:
so are people going to deploy it via L1? 

I thought it would be simpler for them to deploy on L2 directly? (or do we have use cases where they must deploy via L1 message)?

Reply:
I'm not sure how it would work with custom base tokens, whether it's possible to get funds on L2 without a bridge to pay for gas to deploy the bridge
but it's just a standard approach to deploy it via L1, so I decided to stick to it